iavl/iavl.py

```diff
@@ -0,0 +1,392 @@
+    ):
+        """
+        clone a branch node and modify
+        """
+        res = cls(
+            height=node.height,
+            size=node.size,
+            version=version,
+            key=node.key,
+            left_node_ref=node.left_node_ref,
+            right_node_ref=node.right_node_ref,
+        )
+
+        # override
+        if left_node_ref is not None:
+            res.left_node_ref = left_node_ref
+        if right_node_ref is not None:
+            res.right_node_ref = right_node_ref
+        return res
+
+    def persisted(self) -> PersistedNode:
+        return PersistedNode(
+            height=self.height,
+            size=self.size,
+            version=self.version,
+            key=self.key,
+            value=self.value,
+            left_node_ref=self.left_node_ref,
+            right_node_ref=self.right_node_ref,
+        )
+
+    def is_leaf(self):
+        return self.height == 0
+
+    def left_node(self, ndb: NodeDB):
+        return ndb.resolve_node(self.left_node_ref)
+
+    def right_node(self, ndb: NodeDB):
+        return ndb.resolve_node(self.right_node_ref)
+
+    def hash(self):
+        "compute hash of this node"
+        chunks = [
+            cprotobuf.encode_primitive("sint64", self.height),
+            cprotobuf.encode_primitive("sint64", self.size),
+            cprotobuf.encode_primitive("sint64", self.version),
+        ]
+        if self.is_leaf():
+            chunks += encode_bytes(self.key) + encode_bytes(
+                hashlib.sha256(self.value).digest()
+            )
+        else:
+            chunks += encode_bytes(self.left_node_ref) + encode_bytes(
+                self.right_node_ref
+            )
+        return hashlib.sha256(b"".join(chunks)).digest()
+
+    def rotate_left(self, ndb: NodeDB, version: int) -> "Node":
+        r"""
+          S              R
+         / \     =>     / \
+             R         S
+            / \       / \
+          RL             RL
+        """
+        rnode = self.right_node(ndb)
+        self.right_node_ref = rnode.left_node_ref
+        new = Node.from_branch_node(rnode, version, left_node_ref=self)
+        self.update_height_size(ndb)
+        new.update_height_size(ndb)
+        return new
+
+    def rotate_right(self, ndb: NodeDB, version: int):
+        r"""
+           S               L
+          / \      =>     / \
+         L                   S
+        / \                 / \
+          LR               LR
+        """
+        lnode = self.left_node(ndb)
+        self.left_node_ref = lnode.right_node_ref
+        new = Node.from_branch_node(lnode, version, right_node_ref=self)
+        self.update_height_size(ndb)
+        new.update_height_size(ndb)
+        return new
+
+    def update_height_size(self, ndb: NodeDB):
+        lnode = self.left_node(ndb)
+        rnode = self.right_node(ndb)
+        self.height = max(lnode.height, rnode.height) + 1
+        self.size = lnode.size + rnode.size
+
```

Comment:
maybe can align [calcBalance](https://github.com/cosmos/iavl/blob/aff636602ca8dc07a333fae893d49e163d7a7e43/node.go#L508)
```suggestion
    def calc_balance(self, ndb: NodeDB):
        return self.left_node(ndb).height - self.right_node(ndb).height
```

Reply:
done